text/1029-rbac.rst

```diff
@@ -0,0 +1,340 @@
+.. is not in the intersection, the query is rejected without even sending it to
+.. PostgreSQL.
+
+
+When a new role is created, it is assigned no permissions. Roles marked as
+``SUPERUSER`` are deemed to have all permissions.
+
+Since roles are created per-instance, their assigned permissions are applied
+to all branches and their schemas. The ``permissions`` property of roles are
+not name-checked against any schema. They can be set to arbitrary identifiers.
+
+It is required that an instance where roles' permissions are referring to
+permissions in the user schema can have its schema wiped, without affecting
+roles or their assigned permissions. Similarly, it is required that an empty
+instance can be populated with roles and their permissions, before applying
+schema.
+
+When validating, role's permissions are compared to schema permissions by
+their fully-qualified identifiers, converted to strings.
+
+Permissions can only be set by ``SUPERUSER`` users, similarly to how roles can
+only be created and dropped by super users.
+
+
+Per-object access control
+-------------------------
+
+For each created permission, a global with the same name is made available.
+It is evaluated to a boolean value, which is true iff the role of current
+connection has the associated permission.
+
+The global cannot be assigned to, but it is otherwise a normal global variable;
+it can be used in any expression.
+
+In combination with access policies, this allows precise control over which
+objects of an object type are visible or can be modified by each role.
+
+For example::
+
+    CREATE TYPE Secret {
+        CREATE PROPERTY super_secret: str;
+
+        CREATE ACCESS POLICY allow_webapp
+            ALLOW ALL
+            USING (global frontend::webapp);
+
+        CREATE ACCESS POLICY allow_warehouse
+            ALLOW SELECT
+            USING (global data_export);
+
+        CREATE ACCESS POLICY allow_inserts
+            ALLOW INSERT;
+    };
+
+An important detail is that access policies for ``select`` only filter data
+and do not raise errors when an object type is accessed without required
+permissions. It would be more user-friendly to raise an error instead of
+returning no results, but we currently do not provide a reliable way of
+raising errors within access policies. See rejected ideas for justification.
+
+Per-function access control
+---------------------------
+
+Functions cannot have access policies, but we do want to be able to
+limit access to certain functions, such as ``sys::reset_query_stats`` and
+``net::http::schedule_request``.
+
+To provide access control to functions, we introduce a new function metadata
+field named ``require_permission``. It adds the provided permission to the list
+of required permissions of the query, similar to how `DDL` and other existing
+capabilities work.
+
+code::
+
+    CREATE FUNCTION sys::reset_query_stats(...) -> ... {
+        SET volatility := ...
+        SET require_permission := sys::query_stats;
```

Comment:
Should this take a set?

(We may need to fiddle with the syntax here from an implementation perspective; right now, `SET` things get evaluated by the static evaluator, which I think will reject this. We'll need to figure out how we want to do it.)

Reply:
I don't think set would be needed anywhere now, but implementation-wise it is free. So yes, it could be a set, to make it future-proof.